publisher/upload/upload.go

```diff
@@ -0,0 +1,399 @@
+	resp, err := http.Get(u.String())
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode == http.StatusNotFound {
+		l.Printf("[X] Mirroring skipped since since %s was not present in mounted repo", u.String())
+		return nil
+	}
+
+	l.Printf("[ ] Mirror create APT repo %s for %s/%s from %s", srcPath, osVersion, arch, repoUrl)
+	if err = utils.ExecLogOutput(l, "aptly", commandTimeout, "mirror", "create", "-keyring", conf.GpgKeyRing, "mirror-"+osVersion, repoUrl, osVersion, "main"); err != nil {
+		return err
+	}
+	l.Printf("[✔] Mirror create succesfully APT repo %s for %s/%s", srcPath, osVersion, arch)
+
+	l.Printf("[ ] Mirror update APT repo %s for %s/%s", srcPath, osVersion, arch)
+	if err = utils.ExecLogOutput(l, "aptly", commandTimeout, "mirror", "update", "-keyring", conf.GpgKeyRing, "mirror-"+osVersion); err != nil {
+		return err
+	}
+	l.Printf("[✔] Mirror update succesfully APT repo %s for %s/%s", srcPath, osVersion, arch)
+
+	// The last parameter is `Name` that means a query matches all the packages (as it means “package name is not empty”).
+	l.Printf("[ ] Mirror repo import APT repo %s for %s/%s", srcPath, osVersion, arch)
+	if err = utils.ExecLogOutput(l, "aptly", commandTimeout, "repo", "import", "mirror-"+osVersion, osVersion, "Name"); err != nil {
+		return err
+	}
+	l.Printf("[✔] Mirror repo import succesfully APT repo %s for %s/%s", srcPath, osVersion, arch)
+
+	return nil
+}
+
+func uploadFileArtifact(conf config.Config, schema config.UploadArtifactSchema, upload config.Upload, arch string) (err error) {
+	srcPath, destPath := replaceSrcDestTemplates(
+		schema.Src,
+		upload.Dest,
+		conf.RepoName,
+		conf.AppName,
+		arch,
+		conf.Tag,
+		conf.Version,
+		conf.DestPrefix,
+		"")
+
+	srcPath = path.Join(conf.ArtifactsSrcFolder, srcPath)
+	destPath = path.Join(conf.ArtifactsDestFolder, destPath)
+
+	err = utils.CopyFile(srcPath, destPath, upload.Override)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func generateDestinationAssetsPath(downloadedFileName, destPathTemplate, repoName, appName, arch, tag, version, destPrefix, osVersion string) string {
+	destPath := utils.ReplacePlaceholders(destPathTemplate, repoName, appName, arch, tag, version, destPrefix, osVersion)
+	return strings.Replace(destPath, utils.PlaceholderForSrc, downloadedFileName, -1)
+}
+
+func generateAptSrcRepoUrl(template, accessPointHost string) (url string) {
+	url = strings.Replace(template, utils.PlaceholderForAccessPointHost, accessPointHost, -1)
+
+	return
+}
+
+func generateRepoFileContent(accessPointHost, destPath string) (repoFileContent string) {
+
+	contentTemplate := `[newrelic-infra]
+name=New Relic Infrastructure
+baseurl=%s/%s
+gpgkey=https://download.newrelic.com/infrastructure_agent/gpg/newrelic-infra.gpg
+gpgcheck=1
+repo_gpgcheck=1`
+
+	repoFileContent = fmt.Sprintf(contentTemplate, accessPointHost, destPath)
+
+	return
+}
+
+// @TODO: deprecate this function and use generateDestinationAssetsPath() generateDownloadFileName()
```

Comment:
should we take the chance?

Reply:
This PR was mostly to decouple things in packages and iterate over later